### PR TITLE
Dynamic library handles and persistent memory fixes

### PR DESCRIPTION
**Pull request by tbennun**
- [x] ~~Refactor compiled SDFG and invocation~~ (moved to #429)
- [x] Fix CUDA initialization code to be stateless
- [x] Have libraries return an opaque handle
- [x] Have exit function receive only said handle
- [x] Handle instrumentation reports in library handle
- [x] ~~Propagate handle to and~~ use in targets:
    - [x] GPU - CUDA streams and events
    - [x] FPGA - existing FPGA context object
    - [x] MPI - communicator object
- [x] Propagate handle to nested SDFGs as necessary
- [x] Fix persistent memory allocation issues (Issue #324)
- [x] Use `add_global` (With #500 we added a new method to DefinedMemlet, to add "global vars" visible from any function.
The method is add_global.)
- [x] Have init function receive only free symbols and scalars
- [x] Pass library handles (e.g., CUBLAS) to state
- [x] Persistent/global memory allocation and referencing
    - [x] Use declaration/allocation stream in `allocate_array`
    - [x] Use declaration/allocation stream in calls to `dispatch_allocate`
    - [x] Use `__state->` when referring to persistent arrays

### CODE REVIEW

**Comment by TizianoDeMatteis** [comment]
With #500 we added a new method to DefinedMemlet, to add "global vars" visible from any function.
The method is `add_global`.

File: dace/codegen/dispatcher.py [hunk 1]
```diff
@@ -394,17 +399,29 @@ def dispatch_scope(self, map_schedule, sdfg, sub_dfg, state_id,
...
-                          callsite_stream):
+                          allocation_stream):
         """ Dispatches a code generator for data allocation. """
 
         nodedesc = node.desc(sdfg)
         storage = (nodedesc.storage if not isinstance(node, nodes.Tasklet) else
                    dtypes.StorageType.Register)
         self._used_targets.add(self._array_dispatchers[storage])
 
+        # TODO: Move to central allocator
```
Comment (TizianoDeMatteis) [inline dace/codegen/dispatcher.py:410]: OT: Is this a future work?

File: dace/libraries/blas/include/dace_cublas.h [hunk 2]
```diff
@@ -134,18 +129,28 @@ class _CublasConstants {
   double* double_pone_;
   cuComplex* complex64_pone_;
   cuDoubleComplex* complex128_pone_;
+  void* custom_alpha_;
+  void* custom_beta_;
 };
 
-class CublasConstants {
+class CublasHandle {
```
Comment (TizianoDeMatteis) [inline dace/libraries/blas/include/dace_cublas.h:136]: Can you pls add a small comment about the purpose of this class?

**Review by TizianoDeMatteis — CHANGES_REQUESTED** [review]
Looks good to me!
I just have a minor request for adding a  comment.

(Thanks for adding the type hints)

File: dace/codegen/dispatcher.py (hunk 1 above)
Comment (tbennun) [reply dace/codegen/dispatcher.py:410]: yep :) #434

Comment (tbennun) [reply dace/codegen/dispatcher.py:410]: I'll add a comment

File: dace/libraries/blas/include/dace_cublas.h (hunk 2 above)
Comment (tbennun) [reply dace/libraries/blas/include/dace_cublas.h:143]: Done.

**Review by TizianoDeMatteis — APPROVED** [review]

**Pull request merged**